server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -0,0 +1,752 @@
+            assertThat(primaryShard.currentNodeId(), is(equalTo(targetDiscoveryNode.getId())));
+            assertThat(primaryShard.started(), is(equalTo(true)));
+        });
+
+        ensureGreen(indexName);
+        assertDocumentsAreEqual(indexName, numDocs);
+
+        RecoveryState recoveryState = getLatestPeerRecoveryStateForShard(indexName, 0);
+        assertPeerRecoveryWasSuccessful(recoveryState, sourceNode, targetNode);
+
+        long snapshotSizeForIndex = getSnapshotSizeForIndex(repoName, "snap", indexName);
+
+        InstrumentedRepo targetRepo = getRepositoryOnNode(repoName, targetNode);
+        assertThat(targetRepo.totalBytesRead.get(), is(greaterThan(0L)));
+        assertThat(targetRepo.totalBytesRead.get(), is(lessThanOrEqualTo(snapshotSizeForIndex)));
+    }
+
+    private long getSnapshotSizeForIndex(String repository, String snapshot, String index) {
+        GetSnapshotsResponse getSnapshotsResponse =
+            client().admin().cluster().prepareGetSnapshots(repository).addSnapshots(snapshot).get();
+        for (SnapshotInfo snapshotInfo : getSnapshotsResponse.getSnapshots()) {
+            SnapshotInfo.IndexSnapshotDetails indexSnapshotDetails = snapshotInfo.indexSnapshotDetails().get(index);
+            assertThat(indexSnapshotDetails, is(notNullValue()));
+            return indexSnapshotDetails.getSize().getBytes();
+        }
+
+        return -1;
+    }
+
+    private void indexDocs(String indexName, int docIdOffset, int docCount) throws Exception {
+        IndexRequestBuilder[] builders = new IndexRequestBuilder[docCount];
+        BulkRequestBuilder bulkRequestBuilder = client().prepareBulk();
+        for (int i = 0; i < builders.length; i++) {
+            int docId = i + docIdOffset;
+            bulkRequestBuilder.add(
+                client().prepareIndex(indexName)
+                    .setId(Integer.toString(docId))
+                    .setSource("field", docId, "field2", "Some text " + docId)
+            );
+        }
+        BulkResponse bulkResponse = bulkRequestBuilder.get();
+        assertThat(bulkResponse.hasFailures(), is(equalTo(false)));
+
+        // Ensure that the safe commit == latest commit
+        assertBusy(() -> {
+            ShardStats stats = client().admin().indices().prepareStats(indexName).clear().get()
+                .asMap().entrySet().stream().filter(e -> e.getKey().shardId().getId() == 0)
+                .map(Map.Entry::getValue).findFirst().orElse(null);
+            assertThat(stats, is(notNullValue()));
+            assertThat(stats.getSeqNoStats(), is(notNullValue()));
+
+            assertThat(Strings.toString(stats.getSeqNoStats()),
+                stats.getSeqNoStats().getMaxSeqNo(), equalTo(stats.getSeqNoStats().getGlobalCheckpoint()));
+        }, 60, TimeUnit.SECONDS);
+
+        FlushResponse flushResponse = client().admin().indices().prepareFlush(indexName).setForce(true).setWaitIfOngoing(true).get();
```

Comment:
Now that I think about this, once we ensure that the safe commit == latest commit we can remove this flush since the snapshot would do the flush, right?

Reply:
Yeah, that is right.

Perhaps we can then just randomly do the flush before waiting for GCP to advance?

Reply:
Yep, that sounds reasonable to me.

Reply:
I've been thinking about this and I think we cannot flush while there are in-flight operations. This could cause the snapshot to use a commit != safe commit, as it is possible that when we take the snapshot the engine decides that it doesn't need to flush and just uses the latest commit.

Reply:
I wonder if we should advance the safecommit when GCP advances, but let us keep this as is, I think that is orthogonal to this PR.

Reply:
Just circling back here that we do advance the safe commit when GCP advances (thanks Nhat), it is done here: https://github.com/elastic/elasticsearch/blob/58eefdd7575de08ab59ad391e82140c7f915d310/server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java#L89-L94

So it is a timing matter that we could look into resolving in this test in a follow-up.